File: android/src/main/java/com/margelo/nitro/rive/HybridRiveFileFactory.kt
package com.margelo.nitro.rive

import android.annotation.SuppressLint
import androidx.annotation.Keep
import app.rive.runtime.kotlin.core.File
import app.rive.runtime.kotlin.core.Rive
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URI

data class FileAndCache(
  val file: File,
  val cache: ReferencedAssetCache,
  val loader: ReferencedAssetLoader?
)

@Keep
@DoNotStrip
class HybridRiveFileFactory : HybridRiveFileFactorySpec() {
  private fun buildRiveFile(
    data: ByteArray,
    referencedAssets: ReferencedAssetsType?
  ): FileAndCache {
    val cache = mutableMapOf<String, app.rive.runtime.kotlin.core.FileAsset>()
    val loader = ReferencedAssetLoader()
    val customLoader = loader.createCustomLoader(referencedAssets, cache)

    // TODO: The File object in Android does not have the concept of loading CDN assets
    val riveFile = if (customLoader != null) {
      File(data, Rive.defaultRendererType, customLoader)
    } else {
      File(data)
    }

    return FileAndCache(riveFile, cache, if (customLoader != null) loader else null)
  }

  override fun fromURL(url: String, loadCdn: Boolean, referencedAssets: ReferencedAssetsType?): Promise<HybridRiveFileSpec> {
    return Promise.async {
      try {
        val fileAndCache = withContext(Dispatchers.IO) {
          val riveData = HTTPDataLoader.downloadBytes(url)
          buildRiveFile(riveData, referencedAssets)
        }

        val hybridRiveFile = HybridRiveFile()
        hybridRiveFile.riveFile = fileAndCache.file
        hybridRiveFile.referencedAssetCache = fileAndCache.cache
        hybridRiveFile.assetLoader = fileAndCache.loader
        hybridRiveFile
      } catch (e: Exception) {
        throw Error("Failed to download Rive file: ${e.message}", e)
      }
    }
  }

  override fun fromFileURL(fileURL: String, loadCdn: Boolean, referencedAssets: ReferencedAssetsType?): Promise<HybridRiveFileSpec> {
    if (!fileURL.startsWith("file://")) {
      throw Error("fromFileURL: URL must be a file URL: $fileURL")
    }

    return Promise.async {
      try {
        val uri = URI(fileURL)
        val path = uri.path ?: throw Error("fromFileURL: Invalid URL: $fileURL")

        val fileAndCache = withContext(Dispatchers.IO) {
          val riveData = FileDataLoader.loadBytes(path)
          buildRiveFile(riveData, referencedAssets)
        }

        val hybridRiveFile = HybridRiveFile()
        hybridRiveFile.riveFile = fileAndCache.file
        hybridRiveFile.referencedAssetCache = fileAndCache.cache
        hybridRiveFile.assetLoader = fileAndCache.loader
        hybridRiveFile
      } catch (e: Exception) {
        throw Error("Failed to load Rive file: ${e.message}")
      }
    }
  }

  @SuppressLint("DiscouragedApi")
  override fun fromResource(resource: String, loadCdn: Boolean, referencedAssets: ReferencedAssetsType?): Promise<HybridRiveFileSpec> {
    return Promise.async {
      try {
        val fileAndCache = withContext(Dispatchers.IO) {
          val riveData = ResourceDataLoader.loadBytes(resource)
          buildRiveFile(riveData, referencedAssets)
        }

        val hybridRiveFile = HybridRiveFile()
        hybridRiveFile.riveFile = fileAndCache.file
        hybridRiveFile.referencedAssetCache = fileAndCache.cache
        hybridRiveFile.assetLoader = fileAndCache.loader
        hybridRiveFile
      } catch (e: Exception) {
        throw Error("Failed to load Rive file: ${e.message}")
      }
    }
  }

  override fun fromBytes(bytes: ArrayBuffer, loadCdn: Boolean, referencedAssets: ReferencedAssetsType?): Promise<HybridRiveFileSpec> {
    val buffer = bytes.getBuffer(false)
    return Promise.async {
      try {
        val byteArray = ByteArray(buffer.remaining())
        buffer.get(byteArray)
        val fileAndCache = buildRiveFile(byteArray, referencedAssets)
        val hybridRiveFile = HybridRiveFile()
        hybridRiveFile.riveFile = fileAndCache.file
        hybridRiveFile.referencedAssetCache = fileAndCache.cache
        hybridRiveFile.assetLoader = fileAndCache.loader
        hybridRiveFile
      } catch (e: Exception) {
        throw Error("Failed to load Rive file from bytes: ${e.message}")
      }
    }
  }
}

File: ios/HybridRiveFileFactory.swift
import NitroModules
import RiveRuntime

final class HybridRiveFileFactory: HybridRiveFileFactorySpec, @unchecked Sendable {
  let assetLoader = ReferencedAssetLoader()

  /// Asynchronously creates a `HybridRiveFileSpec` by performing the following steps:
  /// 1. Executes `check()` to validate or fetch initial data.
  /// 2. Processes the result with `prepare()`.
  /// 3. If a custom asset loader is available, loads the file using `fileWithCustomAssetLoader(prepared, assetLoader)`.
  ///    Otherwise, loads the file using `file(prepared)`.
  /// 4. Handles referenced assets and caches as needed.
  /// - Parameters:
  ///   - check: Closure to validate or fetch initial data.
  ///   - prepare: Closure to process the checked result.
  ///   - fileWithCustomAssetLoader: Closure to load the file with a custom asset loader.
  ///   - file: Closure to load the file without a custom asset loader.
  ///   - referencedAssets: Optional referenced assets.
  /// - Returns: A promise resolving to a `HybridRiveFileSpec`.
  /// - Throws: Runtime errors if any step fails.
  func genericFrom<CheckResult, Prepared>(
    check: @escaping () throws -> CheckResult,
    prepare: @escaping (CheckResult) throws -> Prepared,
    fileWithCustomAssetLoader: @escaping (Prepared, @escaping LoadAsset) throws -> RiveFile,
    file: @escaping (Prepared) throws -> RiveFile,
    referencedAssets: ReferencedAssetsType?
  ) throws -> Promise<(any HybridRiveFileSpec)> {
    return Promise.async {
      do {
        let checked = try check()

        let result = try await withCheckedThrowingContinuation { continuation in
          DispatchQueue.global(qos: .userInitiated).async {
            do {
              let prepared = try prepare(checked)

              let referencedAssetCache = SendableRef(ReferencedAssetCache())
              let factoryCache: SendableRef<RiveFactory?> = .init(nil)
              let customLoader = self.assetLoader.createCustomLoader(
                referencedAssets: referencedAssets, cache: referencedAssetCache,
                factory: factoryCache)

              let riveFile =
                if let customLoader = customLoader {
                  try fileWithCustomAssetLoader(prepared, customLoader)
                } else {
                  try file(prepared)
                }

              let result = (
                file: riveFile, cache: referencedAssetCache.value, factory: factoryCache.value,
                loader: customLoader != nil ? self.assetLoader : nil
              )
              DispatchQueue.main.async {
                continuation.resume(returning: result)
              }
            } catch {
              DispatchQueue.main.async {
                continuation.resume(throwing: error)
              }
            }
          }
        }

        let hybridRiveFile = HybridRiveFile()
        hybridRiveFile.riveFile = result.file
        hybridRiveFile.referencedAssetCache = result.cache
        if let factory = result.factory {
          hybridRiveFile.cachedFactory = factory
        }
        hybridRiveFile.assetLoader = result.loader
        return hybridRiveFile
      } catch let error as NSError {
        throw RuntimeError.error(
          withMessage: "Failed to download Rive file: \(error.localizedDescription)")
      } catch {
        throw RuntimeError.error(withMessage: "Unknown error occurred while downloading Rive file")
      }
    }
  }

  // MARK: Public Methods
  func fromURL(url: String, loadCdn: Bool, referencedAssets: ReferencedAssetsType?) throws
    -> Promise<(any HybridRiveFileSpec)>
  {
    return try genericFrom(
      check: {
        guard let url = URL(string: url) else {
          throw RuntimeError.error(withMessage: "Invalid URL: \(url)")
        }
        return url
      },
      prepare: { url in try Data(contentsOf: url) },
      fileWithCustomAssetLoader: { (data, loader) in
        try RiveFile(data: data, loadCdn: loadCdn, customAssetLoader: loader)
      },
      file: { (data) in try RiveFile(data: data, loadCdn: loadCdn) },
      referencedAssets: referencedAssets
    )
  }

  func fromFileURL(fileURL: String, loadCdn: Bool, referencedAssets: ReferencedAssetsType?) throws
    -> Promise<(any HybridRiveFileSpec)>
  {
    return try genericFrom(
      check: {
        guard let url = URL(string: fileURL) else {
          throw RuntimeError.error(withMessage: "Invalid URL: \(fileURL)")
        }
        guard url.isFileURL else {
          throw RuntimeError.error(withMessage: "fromFileURL: URL must be a file URL: \(fileURL)")
        }
        return url
      },
      prepare: { url in try Data(contentsOf: url) },
      fileWithCustomAssetLoader: { (data, loader) in
        try RiveFile(data: data, loadCdn: loadCdn, customAssetLoader: loader)
      },
      file: { (data) in try RiveFile(data: data, loadCdn: loadCdn) },
      referencedAssets: referencedAssets
    )
  }

  func fromResource(resource: String, loadCdn: Bool, referencedAssets: ReferencedAssetsType?) throws
    -> Promise<(any HybridRiveFileSpec)>
  {
    return try genericFrom(
      check: {
        guard Bundle.main.path(forResource: resource, ofType: "riv") != nil else {
          throw RuntimeError.error(withMessage: "Could not find Rive file: \(resource).riv")
        }
        return resource
      },
      prepare: { $0 },
      fileWithCustomAssetLoader: { (resource, loader) in
        try RiveFile(resource: resource, loadCdn: loadCdn, customAssetLoader: loader)
      },
      file: { (resource) in try RiveFile(resource: resource, loadCdn: loadCdn) },
      referencedAssets: referencedAssets
    )
  }

  func fromResource(resource: String, loadCdn: Bool) throws -> Promise<(any HybridRiveFileSpec)> {
    return try fromResource(resource: resource, loadCdn: loadCdn, referencedAssets: nil)
  }

  func fromBytes(bytes: ArrayBufferHolder, loadCdn: Bool, referencedAssets: ReferencedAssetsType?)
    throws -> Promise<
      (any HybridRiveFileSpec)
    >
  {
    return try genericFrom(
      check: { bytes.toData(copyIfNeeded: false) },
      prepare: { $0 },
      fileWithCustomAssetLoader: { (data, loader) in
        try RiveFile(data: data, loadCdn: loadCdn, customAssetLoader: loader)
      },
      file: { (data) in try RiveFile(data: data, loadCdn: loadCdn) },
      referencedAssets: referencedAssets
    )
  }
}
